rest_server/IOMgrConnection.go

```diff
@@ -0,0 +1,444 @@
+	JSONPayloadBuffer []byte, size int) (bytesWritten int, err error) {
+
+	// Write Payload length
+	var buf [binary.MaxVarintLen32]byte
+	encodedLength := binary.PutUvarint(buf[:], uint64(size))
+	if _, err = IOMgrConn.Write(buf[:encodedLength]); err != nil {
+		return 0, err
+	}
+	siodbLoggerPool.Debug("writeJSONPayloadChunk | Payload size written to IOMgr: %v",
+		uint64(size))
+
+	if size > 0 {
+		// Write raw payload
+		if bytesWritten, err = IOMgrConn.Write(JSONPayloadBuffer[:size]); err != nil {
+			siodbLoggerPool.Error("err: %v", err)
+			return bytesWritten, fmt.Errorf("not able to write payload to IOMgr: %v", err)
+		}
+	}
+
+	return bytesWritten, nil
+}
+
+func (IOMgrConn *IOMgrConnection) writeIOMgrRequest(
+	restVerb SiodbIomgrProtocol.RestVerb,
+	objectType SiodbIomgrProtocol.DatabaseObjectType,
+	userName string,
+	token string,
+	objectName string,
+	objectId uint64) (requestId uint64, err error) {
+
+	requestId = IOMgrConn.RequestID
+	IOMgrConn.RequestID++
+	var databaseEngineRestRequest SiodbIomgrProtocol.DatabaseEngineRestRequest
+	databaseEngineRestRequest.RequestId = requestId
+	databaseEngineRestRequest.Verb = restVerb
+	databaseEngineRestRequest.ObjectType = objectType
+	databaseEngineRestRequest.UserName = userName
+	databaseEngineRestRequest.Token = token
+
+	if len(objectName) > 0 {
+		databaseEngineRestRequest.ObjectName = objectName
+	}
+	if objectId > 0 {
+		databaseEngineRestRequest.ObjectId = objectId
+	}
+	siodbLoggerPool.Debug("writeIOMgrRequest | databaseEngineRestRequest: %v", databaseEngineRestRequest)
+	siodbLoggerPool.Debug("writeIOMgrRequest | IOMgrConn.RequestID: %v", requestId)
+	siodbLoggerPool.Debug("writeIOMgrRequest | IOMgrConn.RequestID++: %v", IOMgrConn.RequestID)
+
+	if _, err := IOMgrConn.writeMessage(DATABASEENGINERESTREQUEST, &databaseEngineRestRequest); err != nil {
+		return requestId, fmt.Errorf("Unable to write message to IOMgr: %v", err)
+	}
+
+	return requestId, nil
+}
+
+func (IOMgrConn *IOMgrConnection) readIOMgrResponse(requestID uint64) (err error) {
+
+	var databaseEngineResponse SiodbIomgrProtocol.DatabaseEngineResponse
+
+	if _, err := IOMgrConn.readMessage(DATABASEENGINERESPONSE, &databaseEngineResponse); err != nil {
+		return fmt.Errorf("Unable to read response from IOMgr: %v", err)
+	}
+	siodbLoggerPool.Debug("readIOMgrResponse | databaseEngineResponse: %v", databaseEngineResponse)
+	siodbLoggerPool.Debug("readIOMgrResponse | databaseEngineResponse.RequestId: %v", databaseEngineResponse.RequestId)
+	siodbLoggerPool.Debug("readIOMgrResponse | IOMgrConn.RequestID: %v", IOMgrConn.RequestID)
+
+	if databaseEngineResponse.RequestId != requestID {
+		return fmt.Errorf("request IDs mismatch")
+	}
+
+	if len(databaseEngineResponse.Message) > 0 {
+		return fmt.Errorf("code: %v, message: %v", databaseEngineResponse.Message[0].GetStatusCode(), databaseEngineResponse.Message[0].GetText())
+	}
+
+	return nil
+}
+
+func (IOMgrConn *IOMgrConnection) readChunkedJSON(c *gin.Context) (err error) {
```

Comment:
This function got too complicated.
Here's in pseudocode what is really required:
```c++
httpChunkBuffer = new byte[configuration.httpChunkSize]

const uint32 MAX_IOMGR_CHUNKED_PROTOCOL_CHUNK_SIZE = 2^31 - 1

uint32 httpBufferPos = 0

while (true) {
	uint32 chunkSize = readChunkSizeFromIOMgr();
	checkReadError()
	
	if (chunkSize == 0) break
	if (chunkSize > MAX_IOMGR_CHUNKED_PROTOCOL_CHUNK_SIZE)
		return ProtocolError
	
	while (chunkSize > 0) {
		maxReadSize = min(httpChunkBuffer.length - httpBufferPos, chunkSize)
		readBytes = readBytesFromIomgr(
                 into httpChunkBuffer starting at position httpBufferPos, maxReadSize);
		checkReadError, including if read less than maxReadSize

		chunkSize -= readSize
		httpBufferPos += readSize

		if (httpBufferPos == httpChunkBuffer.length) {
			writeHttpChunkOfSize(httpChunkBuffer, httpChunkBuffer.length);
			checkWriteError()
			httpBufferPos = 0
		}
	}
}

if (httpBufferPos > 0) {
	writeHttpChunkOfSize(httpChunkBuffer, httpBufferPos);
	checkWriteError()
}
```